Remotion/Data/DomainObjects/Persistence/ReadOnlyStorageProviderDecorator.cs

```diff
@@ -0,0 +1,118 @@
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using Remotion.Data.DomainObjects.DataManagement;
+using Remotion.Data.DomainObjects.Mapping;
+using Remotion.Data.DomainObjects.Mapping.SortExpressions;
+using Remotion.Data.DomainObjects.Queries;
+using Remotion.Data.DomainObjects.Queries.Configuration;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence
+{
+  /// <summary>
+  /// Decorator for <see cref="IReadOnlyStorageProvider"/> to ensure that <see cref="IQuery"/>s have the appropriate <see cref="QueryType"/> when
+  /// being executed in the ReadOnly context.
+  /// </summary>
+  public class ReadOnlyStorageProviderDecorator : IReadOnlyStorageProvider
+  {
+    public ReadOnlyStorageProviderDecorator (IReadOnlyStorageProvider innerStorageProvider)
+    {
+      ArgumentUtility.CheckNotNull("innerStorageProvider", innerStorageProvider);
+
+      InnerStorageProvider = innerStorageProvider;
+    }
+
+    public IReadOnlyStorageProvider InnerStorageProvider { get; }
+
+    public void Dispose ()
+    {
+      InnerStorageProvider.Dispose();
+    }
+
+    public ObjectLookupResult<DataContainer> LoadDataContainer (ObjectID id)
+    {
+      ArgumentUtility.CheckNotNull("id", id);
+
+      return InnerStorageProvider.LoadDataContainer(id);
+    }
+
+    public IEnumerable<ObjectLookupResult<DataContainer>> LoadDataContainers (IReadOnlyCollection<ObjectID> ids)
+    {
+      ArgumentUtility.CheckNotNull("ids", ids);
+
+      return InnerStorageProvider.LoadDataContainers(ids);
+    }
+
+    public IEnumerable<DataContainer> LoadDataContainersByRelatedID (
+        RelationEndPointDefinition relationEndPointDefinition,
+        SortExpressionDefinition? sortExpressionDefinition,
+        ObjectID relatedID)
+    {
+      ArgumentUtility.CheckNotNull("relationEndPointDefinition", relationEndPointDefinition);
+      ArgumentUtility.CheckNotNull("relatedID", relatedID);
+
+      return InnerStorageProvider.LoadDataContainersByRelatedID(relationEndPointDefinition, sortExpressionDefinition, relatedID);
+    }
+
+    public IEnumerable<DataContainer?> ExecuteCollectionQuery (IQuery query)
+    {
+      ArgumentUtility.CheckNotNull("query", query);
+
+      if (query.QueryType != QueryType.CollectionReadOnly)
+        throw new ArgumentException($"Query '{query}' with query type '{query.QueryType}' cannot be used with an IReadOnlyStorageProvider.ExecuteCollectionQuery.", nameof(query));
+
+      return InnerStorageProvider.ExecuteCollectionQuery(query);
+    }
+
+    public IEnumerable<IQueryResultRow> ExecuteCustomQuery (IQuery query)
+    {
+      ArgumentUtility.CheckNotNull("query", query);
+
+      if (query.QueryType != QueryType.CustomReadOnly)
+        throw new ArgumentException($"Query '{query}' with query type '{query.QueryType}' cannot be used with an IReadOnlyStorageProvider.ExecuteCustomQuery.", nameof(query));
```

Comment:
The decorator should only throw when the QueryType is read-write; the decorator is not responsible for detecting the wrong result type - that is done by the decorated StorageProvider.
Use a Check-method to throw when the QueryType is read-write, and use that in all three Execute methods.
Update the unit tests accordingly.